### DISCUSSION TITLE
raw query with parameter not working on DDL query

### DISCUSSION THREAD

**Discussion by BimaAdi** [sqlalchemy/sqlalchemy#10402] [Usage Questions] [answered]
### Describe the bug

I try to switch database driver from psycopg2 to psycopg but I found some odd error behavior when performing raw DDL query with parameter.

### SQLAlchemy Version in Use

2.0.21

### DBAPI (i.e. the database driver)

psycopg

### Database Vendor and Major Version

PostgreSQL 14

### Python Version

3.10.10

### Operating system

Linux

### To Reproduce

```python
from sqlalchemy import Boolean, Column, Integer, String, create_engine, text
from sqlalchemy.orm import sessionmaker, scoped_session, declarative_base

# Create sqlalchemy session
username = "postgres"
password = "password"
host = "127.0.0.1"
port = 5432
database = "postgres"

engine_psycopg2 = create_engine(
    f"postgresql+psycopg2://{username}:{password}@{host}:{port}/{database}",
    pool_size=20,
    max_overflow=0,
    pool_timeout=300,
)
Session_psycopg2 = sessionmaker(engine_psycopg2, future=True)
db_session_psycopg2 = scoped_session(Session_psycopg2)

engine_psycopg3 = create_engine(
    f"postgresql+psycopg://{username}:{password}@{host}:{port}/{database}",
    pool_size=20,
    max_overflow=0,
    pool_timeout=300,
)
Session_psycopg3 = sessionmaker(engine_psycopg3, future=True)
db_session_psycopg3 = scoped_session(Session_psycopg3)

# base for model
Base = declarative_base()


class Todo(Base):
    __tablename__ = "todo"

    id = Column("id", Integer, primary_key=True, nullable=False, index=True)
    todo = Column("todo", String, nullable=False)
    is_done = Column("is_done", Boolean, default=False)


def create_all_table():
    Base.metadata.create_all(engine_psycopg3)


def delete_all_table():
    Base.metadata.drop_all(engine_psycopg3)


create_all_table()

with db_session_psycopg2() as db:
    db.execute(
        text("ALTER SEQUENCE public.todo_id_seq RESTART :x"),
        {"x": 3},
    )  # working on psycopg2

with db_session_psycopg3() as db:
    db.execute(
        text("ALTER SEQUENCE public.todo_id_seq RESTART :x"),
        {"x": 5},
    )  # not working on psycopg, produce error on stacktrace

delete_all_table()
```

### Error
```
Traceback (most recent call last):
  File ".../env/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File ".../env/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 921, in do_execute
    cursor.execute(statement, parameters)
  File ".../env/lib/python3.10/site-packages/psycopg/cursor.py", line 737, in execute
    raise ex.with_traceback(None)
psycopg.errors.SyntaxError: syntax error at or near "$1"
LINE 1: ALTER SEQUENCE public.todo_id_seq RESTART $1
                                                  ^

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File ".../main.py", line 58, in <module>
    db.execute(
  File ".../env/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2262, in execute
    return self._execute_internal(
  File ".../env/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2153, in _execute_internal
    result = conn.execute(
  File ".../env/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1412, in execute
    return meth(
  File ".../env/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 516, in _execute_on_connection
    return connection._execute_clauseelement(
  File ".../env/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1635, in _execute_clauseelement
    ret = self._execute_context(
  File ".../env/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1844, in _execute_context
    return self._exec_single_context(
  File ".../env/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1984, in _exec_single_context
    self._handle_dbapi_exception(
  File ".../env/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2339, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File ".../env/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File ".../env/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 921, in do_execute
    cursor.execute(statement, parameters)
  File ".../env/lib/python3.10/site-packages/psycopg/cursor.py", line 737, in execute
    raise ex.with_traceback(None)
sqlalchemy.exc.ProgrammingError: (psycopg.errors.SyntaxError) syntax error at or near "$1"
LINE 1: ALTER SEQUENCE public.todo_id_seq RESTART $1
                                                  ^
[SQL: ALTER SEQUENCE public.todo_id_seq RESTART %(x)s]
[parameters: {'x': 5}]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

Since you are using a text query this isn't a behaviour that sqlalchemy controls.

DDL statements don't support parameters in general but it depends on the driver if they do allow them or not.

You should either use literal binds in sqlalchemy or look of psycopg supports doing the same

**Comment by bouillon** [comment]
Interesting, we also use here lot ow raw sql queries and after migration of to psycopg3 it isn't works anymore.
I found only this https://docs.sqlalchemy.org/en/20/core/engines.html#hiding-parameters in official docs.

The problem is we have a lot of queries like

```
   db.execute(
        text("select somethin  where x=:x"),
        {"x": 5},
    ) 
```

What would be the best solution to execute them without modify query?
Using psycopg3 would be a solution, but it has another syntax for parameters.

**Reply by zzzeek** [reply]
a SELECT query supports parameters and there is no need to change them, assuming they are used for literal bound values only.  the question here refers only to DDL queries like ALTER SEQUENCE in spaces that bound values aren't supported by the backend.

**Reply by CaselIT** [reply]
You can add bind parameters to the text using https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.TextClause.bindparams specifying https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.bindparam.params.literal_execute to true.

Another alternative could be to use client cursor in psycopg, but it's not currently implemented in sqlalchemy https://github.com/sqlalchemy/sqlalchemy/discussions/8978 (but it may be possible to implement externally)

**Reply by bouillon** [reply]
Thanks I got it. 
I was not exact enough.  He are sql statements in use with 

```
query="""
 DO $$ BEGIN
..... :alchemy_param1, :alchemy_param2
 END $$;
"""
```

Similar to https://github.com/jackc/pgx/discussions/1652 but parameter format is sqlachemy.

.execute is working with psycopg2 but not psycopg3.

I  understand now why it's not working any more but I'm asking me if this is correct behavior.
I would expect that if i'm using sqlalchemy parameters and pass them to execute method (which is part of sqlalchemy) than the sqlalchemy manage to bind this params on his  high level.

If i'm using connection.execute and not direct connection with psycopg and sqlalchemy parameter format :param1, :param2 i shouldn't care about internals of psycopg3.

This is in my case simply replace the values by
query.replace(":alchemy_param1", value)  (will work for me because i have here only int values)

But also params are not working with json


`myjsonfield @> '{"field": :field}' and INTERVAL '10 days' if 10 days is a parameter `




Correct me if i'm wrong...

**Reply by CaselIT** [reply]
> I would expect that if i'm using sqlalchemy parameters and pass them to execute method (which is part of sqlalchemy) than the sqlalchemy manage to bind this params on his high level.

the parameters are passed to the driver that knows best how to handle them.
No abstraction library should bind them by default, since that may lead to security issues or performance problems.

> If i'm using connection.execute and not direct connection with psycopg and sqlalchemy parameter format :param1, :param2 i shouldn't care about internals of psycopg3.

The parameters are transformed to something psycopg3 can understand but the binding of the parameters is left to the driver.

As I mentioned you can make sqlalchemy do some of the binding by using literal execute. That's better than doing `query.replace(":alchemy_param1", value)`. But in general you should resort to it only if the normal way does not work.

Note that you may also configure psycopg to use ClientCursor https://www.psycopg.org/psycopg3/docs/advanced/cursors.html#client-side-binding-cursors by providing that parameter using connect_args https://docs.sqlalchemy.org/en/20/core/engines.html#use-the-connect-args-dictionary-parameter.
Something like this should work
```py
import sqlalchemy as sa
from psycopg import ClientCursor

sa.create_engine(..., connect_args={"cursor_factory": ClientCursor})
```

**Reply by CaselIT** [reply]
btw: this is already documented https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#using-a-different-cursor-class
